npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
+	// Here Run periodic task to check if any sets with empty references are present and delete them
+}
 
 func (dp *DPShim) GetIPSet(setName string) *ipsets.IPSet {
 	return nil
 }
 
+func (dp *DPShim) getIPSet(setName string) *controlplane.ControllerIPSets {
+	return dp.setCache[setName]
+}
+
+func (dp *DPShim) setExists(setName string) bool {
+	_, ok := dp.setCache[setName]
+	return ok
+}
+
 func (dp *DPShim) CreateIPSets(setNames []*ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
+	for _, set := range setNames {
+		dp.createIPSet(set)
+	}
+}
+
+func (dp *DPShim) createIPSet(set *ipsets.IPSetMetadata) {
+	setName := set.GetPrefixName()
+
+	if dp.setExists(setName) {
+		return
+	}
+
+	dp.setCache[setName] = controlplane.NewControllerIPSets(set)
+	dp.dirtyCache.modifyAddorUpdateSets(setName)
 }
 
 func (dp *DPShim) DeleteIPSet(setMetadata *ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
+	dp.deleteIPSet(setMetadata)
 }
 
-func (dp *DPShim) AddToSets(setNames []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
-	return nil
+func (dp *DPShim) deleteIPSet(setMetadata *ipsets.IPSetMetadata) {
+	setName := setMetadata.GetPrefixName()
+	klog.Infof("deleteIPSet: cleaning up %s", setName)
+	set, ok := dp.setCache[setName]
+	if !ok {
+		return
+	}
+
+	if set.HasReferences() {
+		return
+	}
+
+	delete(dp.setCache, setName)
+	dp.dirtyCache.modifyDeleteSets(setName)
 }
 
-func (dp *DPShim) RemoveFromSets(setNames []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
+func (dp *DPShim) AddToSets(setMetadatas []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
+	dp.Lock()
+	defer dp.Unlock()
+	for _, set := range setMetadatas {
+		prefixedSetName := set.GetPrefixName()
+		if !dp.setExists(prefixedSetName) {
+			dp.createIPSet(set)
+		}
+
+		set := dp.setCache[prefixedSetName]
+		if set.IPSetMetadata.GetSetKind() != ipsets.HashSet {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is not a hash set", prefixedSetName))
+		}
+
+		cachedPod, ok := set.IPPodMetadata[podMetadata.PodIP]
+		set.IPPodMetadata[podMetadata.PodIP] = podMetadata
+		if ok && cachedPod.PodKey != podMetadata.PodKey {
+			klog.Infof("AddToSet: PodOwner has changed for Ip: %s, setName:%s, Old podKey: %s, new podKey: %s. Replace context with new PodOwner.",
+				cachedPod.PodIP, set.Name, cachedPod.PodKey, podMetadata.PodKey)
+			continue
+		}
+
+		dp.dirtyCache.modifyAddorUpdateSets(prefixedSetName)
+	}
+
 	return nil
 }
 
-func (dp *DPShim) AddToLists(listName, setNames []*ipsets.IPSetMetadata) error {
+func (dp *DPShim) RemoveFromSets(setMetadatas []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
+	dp.Lock()
+	defer dp.Unlock()
+	for _, set := range setMetadatas {
+		prefixedSetName := set.GetPrefixName()
+		if !dp.setExists(prefixedSetName) {
+			dp.createIPSet(set)
+		}
+
+		set := dp.setCache[prefixedSetName]
+		if set.IPSetMetadata.GetSetKind() != ipsets.HashSet {
```

Comment:
we should generalize addToSet, deleteFromSet, addToList, ... validations into a dp func for both dpshim and Dataplane